finagle/h2/src/main/scala/com/twitter/finagle/buoyant/h2/StreamProxy.scala

```diff
@@ -1,23 +1,53 @@
 package com.twitter.finagle.buoyant.h2
 import com.twitter.util.{Future, Try}
+import scala.collection.mutable
 
-/**
- * Wraps an underlying [[Stream]] with an [[StreamProxy.onFrame onFrame]] function.
- * The `onFrame` function is called for each frame in the stream.
- * @param underlying the [[Stream]] wrapped by this proxy
- * @param onFrame function called for each [[Frame]] in the underlying [[Stream]]
- */
-class StreamProxy(underlying: Stream, onFrame: Try[Frame] => Unit) extends Stream {
+abstract class StreamProxy(underlying: Stream) extends Stream {
   override def isEmpty: Boolean = underlying.isEmpty
-
-  override def read(): Future[Frame] = underlying.read().respond(onFrame)
-
   /**
    * Satisfied when an end-of-stream frame has been read from this
    * stream.
    *
    * If the stream is reset prematurely, onEnd fails with a [[Reset]].
    */
   override def onEnd: Future[Unit] = underlying.onEnd
+}
+
+/**
+ * Wraps an underlying [[Stream]] with an [[StreamOnFrame.onFrame onFrame]] function.
+ * The `onFrame` function is called for each frame in the stream.
+ *
+ * @param underlying the [[Stream]] wrapped by this proxy
+ * @param onFrame function called for each [[Frame]] in the underlying [[Stream]]
+ */
+// TODO: consider renaming `onFrame` to `foreach`?
+class StreamOnFrame(underlying: Stream, onFrame: Try[Frame] => Unit) extends StreamProxy(underlying) {
+  override def read(): Future[Frame] = underlying.read().respond(onFrame)
   override def toString: String = s"StreamProxy($underlying, onFrame=$onFrame)"
 }
+
+class StreamFlatMap(underlying: Stream, f: Try[Frame] => Seq[Frame]) extends StreamProxy(underlying) {
```

Comment:
I think we want `f` to be of type `Frame => Seq[Frame]`.  In the case where we fail when reading a Frame off of the underlying stream, I think we want to also fail the resulting stream.  

To put this in context: if we read part of a response stream and then hit an error, I think we want to transmit that error to the caller (or retry the whole request).  I don't think we want to tack a Trailers frame onto the end of the incomplete stream.